suc/pkg/host/upgrade.go

```diff
@@ -0,0 +1,123 @@
+		logrus.Debugf("wins.exe is up to date (%s)", currentVersion)
+		return false, nil
+	}
+
+	restartService, upgradeErr := updateBinaries(desiredVersion)
+	if upgradeErr != nil {
+		return false, upgradeErr
+	}
+
+	return restartService, nil
+}
+
+// updateBinaries writes the embedded binary onto the disk in the rancher-wins config directory (c:\etc\rancher\wins, by default).
+// Once written, the binary is invoked to ensure that it is not corrupted and is running the expected version.
+// After confirming the version, the updated binary is moved into the wins.exe binary directory ('c:\usr\local\bin', by default)
+// and 'c:\Windows' directories. Once the upgraded binary has been moved into place, it is invoked once again
+// to confirm the file was copied correctly.
+func updateBinaries(desiredVersion string) (bool, error) {
+	logrus.Info("Writing updated wins.exe to disk")
+	// write the embedded binary to disk
+	updatedBinaryPath := fmt.Sprintf("%s/wins-%s.exe", getWinsConfigDir(), strings.Trim(desiredVersion, "\n"))
+	err := os.WriteFile(updatedBinaryPath, winsBinary, os.ModePerm)
+	if err != nil {
+		return false, err
+	}
+
+	// confirm that the new binary works and returns the version that we expect
+	err = confirmWinsBinaryVersion(desiredVersion, updatedBinaryPath)
+	if err != nil {
+		return false, fmt.Errorf("failed to stage updated binary: %w", err)
+	}
+
+	logrus.Info("Stopping rancher-wins...")
+	rw, _, err := service.OpenRancherWinsService()
+	if err != nil {
+		return false, fmt.Errorf("failed to open rancher-wins service while attempting to upgrade binary: %w", err)
+	}
+
+	// The service needs to be stopped before we can modify
+	// the binary it uses
+	err = rw.Stop()
+	if err != nil {
+		return false, fmt.Errorf("failed to stop rancher-wins service while attempting to upgrade binary: %w", err)
+	}
+
+	logrus.Infof("Copying %s to %s", updatedBinaryPath, defaultWinsPath)
+	err = copyFile(updatedBinaryPath, defaultWinsPath)
+	if err != nil {
+		return false, fmt.Errorf("failed to copy new wins.exe binary to %s: %w", defaultWinsPath, err)
+	}
+
+	// While the rancher-wins service looks for wins.exe in c:\Windows
+	// for consistency’s sake we should also ensure it's updated in c:\usr\local\bin
+	// as the install script places it there as well
+	usrLocalBinPath := getWinsUsrLocalBinBinary()
+
+	logrus.Infof("Copying %s to %s", updatedBinaryPath, usrLocalBinPath)
+	err = copyFile(updatedBinaryPath, usrLocalBinPath)
+	if err != nil {
+		return false, fmt.Errorf("failed to copy new wins.exe binary to %s: %w", usrLocalBinPath, err)
+	}
+
+	logrus.Infof("Validating updated binaries...")
+	err = confirmWinsBinaryVersion(desiredVersion, defaultWinsPath)
+	if err != nil {
+		return false, err
+	}
+
+	err = confirmWinsBinaryVersion(desiredVersion, getWinsUsrLocalBinBinary())
+	if err != nil {
+		return false, err
+	}
+
+	logrus.Infof("Removing %s", updatedBinaryPath)
+	err = os.Remove(updatedBinaryPath)
```

Comment:
Why are we removing the source binary here? Could it be possible that the binary at the target location gets removed somehow, and we would need to copy it over again?

Reply:
The intent here was to ensure that we aren't leaving binaries on the host that aren't being used. Do you mean that it would be useful for users to have access to this file in case they need to manually restore to that version? 

Technically there _are_ two copies of wins.exe on each node, so there is still a backup that can be used to manually update the node if one is deleted. lmk if I've misunderstood your comment though 😄

Reply:
I was initially considering the case where the function might be invoked again and fail to find the binary because it was removed. However, I now realize that the binary is recreated each time the function is called, so my question is no longer relevant.